gapis/api/vulkan/framegraph.go

```diff
@@ -16,14 +16,153 @@ package vulkan
+
+	rpInfos := []*renderpassInfo{}
+	var rpInfo *renderpassInfo
+	currRpId := uint64(0)
+
+	// postSubCmdCb effectively processes each subcommand to extract renderpass info.
+	postSubCmdCb := func(state *api.GlobalState, subCmdIdx api.SubCmdIdx, cmd api.Cmd, i interface{}) {
+		vkState := GetState(state)
+		cmdRef := i.(CommandReferenceʳ)
+		cmdArgs := GetCommandArgs(ctx, cmdRef, vkState)
+
+		// Beginning of renderpass
+		if _, ok := cmdArgs.(VkCmdBeginRenderPassArgsʳ); ok {
+			if rpInfo != nil {
+				panic("Renderpass starts without having ended")
+			}
+			rpInfo = &renderpassInfo{
+				id:       currRpId,
+				beginIdx: subCmdIdx,
+				nodes:    []dependencygraph2.NodeID{},
+				deps:     make(map[uint64]struct{}),
+			}
+			currRpId++
+		}
+
+		// Process commands that are inside a renderpass
+		if rpInfo != nil {
+			nodeID := dependencyGraph.GetCmdNodeID(api.CmdID(subCmdIdx[0]), subCmdIdx[1:])
+			rpInfo.nodes = append(rpInfo.nodes, nodeID)
+		}
+
+		// Ending of renderpass
+		if _, ok := cmdArgs.(VkCmdEndRenderPassArgsʳ); ok {
+			if rpInfo == nil {
+				panic("Renderpass ends without having started")
+			}
+			rpInfo.endIdx = subCmdIdx
+			rpInfos = append(rpInfos, rpInfo)
+			rpInfo = nil
+		}
+	}
+
+	// Iterate on the capture commands to gather information
+	c, err := capture.ResolveGraphicsFromPath(ctx, p)
+	if err != nil {
+		return nil, err
+	}
+	if err := sync.MutateWithSubcommands(ctx, p, c.Commands, nil, nil, postSubCmdCb); err != nil {
+		return nil, err
+	}
+
+	updateDependencies(rpInfos, dependencyGraph)
+
+	// Build the framegraph nodes and edges from collected data.
+	nodes := make([]*api.FramegraphNode, len(rpInfos))
+	for i, rpInfo := range rpInfos {
+		// Graphviz DOT: use "\l" as a newline to obtain left-aligned text.
+		text := fmt.Sprintf("Renderpass %v\\lbegin:%v\\lend:%v\\l", rpInfo.id, rpInfo.beginIdx, rpInfo.endIdx)
+		nodes[i] = &api.FramegraphNode{
+			Id:   rpInfo.id,
+			Text: text,
+		}
+	}
+
+	edges := []*api.FramegraphEdge{}
+	for _, rpInfo = range rpInfos {
+		for deps := range rpInfo.deps {
+			edges = append(edges, &api.FramegraphEdge{
+				// We want the graph to show the flow of how the frame is
+				// created (rather than the flow of dependencies), so use the
+				// dependency as the edge origin and rpInfo as the destination.
+				Origin:      deps,
+				Destination: rpInfo.id,
+			})
+		}
+	}
+
+	return &api.Framegraph{Nodes: nodes, Edges: edges}, nil
+}
+
+// updateDependencies establishes dependencies between renderpasses.
+func updateDependencies(rpInfos []*renderpassInfo, dependencyGraph dependencygraph2.DependencyGraph) {
+	// isInsideRenderpass: node -> renderpass it belongs to.
+	isInsideRenderpass := map[dependencygraph2.NodeID]uint64{}
+	for _, rpInfo := range rpInfos {
+		for _, n := range rpInfo.nodes {
+			isInsideRenderpass[n] = rpInfo.id
+		}
+	}
+	// node2renderpasses: node -> set of renderpasses it depends on.
+	node2renderpasses := map[dependencygraph2.NodeID]map[uint64]struct{}{}
```

Comment:
Future comment: This does not do anything for now but if you plan to make type assertion on the `struct{}` that's a lot of type assertion there.

Reply:
This is a canonical way to represent a set in Go: `map[<type of your set>]struct{}`. Then you can use the map to test for presence in set, etc...